File: jscomp/test/gpr_2614_test.ml

[@@@ocaml.warning "-a"]

type t = {

  mutable hi : int
    [@bs.as "Content-Type"];
  mutable low : int
    [@bs.as "l"];
  mutable x : int;
    [@bs.as "open"]
} [@@bs.deriving abstract]
  (* [@@bs.x] *)


let v = t ~hi:3 ~low:2 ~x:2


let (a,b,c) = (v |. hi, v |. low, v |. x)

(**

  v |. (hi, lo)
*)
let ff () =
  v |. hiSet 3;
  v |. lowSet 2


type a = {
  mutable low : string option
      [@bs.as "lo-x"]
;
  hi : int
} [@@bs.deriving abstract]


(**
external a : ?low:int -> hi:int -> a
low: a -> int option [@@bs.return undefined_to_opt]
lowSet : a -> int -> unit
*)
let h0 =
  a ~hi:2 ~low:"x"

let h1 =   a ~hi:2 ~low:"x" ()

let h2 = a ~hi:2 ()

let hh x =
  x |. lowSet "3";
  x |. low

(** should we make the type of

    lowSet : a -> string option -> unit
    lowSet : a -> string -> unit
*)

let hh2 x =
  match x |. low with
  | None -> 0
  | Some _ -> 1

type css =
  {
    a0 : int option ;
    a1 : int option ;
    a2 : int option ;
    a3 : int option ;
    a4 : int option ;
    a5 : int option ;
    a6 : int option ;
    a7 : int option ;
    a8 : int option ;
    a9 : int option
    [@bs.as "xx-yy"];
    a10 : int option ;
    a11 : int option ;
    a12 : int option ;
    a13 : int option ;
    a14 : int option ;
    a15 : int option ;
  }
  [@@bs.deriving abstract]


let u = css ~a9:3 ()
let v =
  match u |. a9 with
  | None -> 0
  | Some x -> x 

File: jscomp/test/gpr_2614_test.js
'use strict';

var Js_primitive = require("../../lib/js/js_primitive.js");

var v = {
  "Content-Type": 3,
  l: 2,
  open: 2
};

v["Content-Type"];

var b = v.l;

var c = v.open;

function ff() {
  v["Content-Type"] = 3;
  v.l = 2;
  return /* () */0;
}

var partial_arg = /* Some */["x"];

function h0() {
  var tmp = {
    hi: 2
  };
  if (partial_arg) {
    tmp["lo-x"] = partial_arg[0];
  }
  return tmp;
}

var h1 = {
  "lo-x": "x",
  hi: 2
};

var h2 = {
  hi: 2
};

function hh(x) {
  x["lo-x"] = "3";
  return Js_primitive.undefined_to_opt(x["lo-x"]);
}

function hh2(x) {
  var match = x["lo-x"];
  if (match !== undefined) {
    return 1;
  } else {
    return 0;
  }
}

var u = {
  "xx-yy": 3
};

var match = u["xx-yy"];

var v$1 = match !== undefined ? match : 0;

exports.b = b;
exports.c = c;
exports.ff = ff;
exports.h0 = h0;
exports.h1 = h1;
exports.h2 = h2;
exports.hh = hh;
exports.hh2 = hh2;
exports.u = u;
exports.v = v$1;
/*  Not a pure module */
